### PR TITLE
HTML-321 - workflowState tag changing state start dates when re-submitting a form

### PR DESCRIPTION
**Pull request by mseaton**

### CODE REVIEW

File: api-tests/src/test/java/org/openmrs/module/htmlformentry/WorkflowStateTagTest.java [hunk 1]
```diff
@@ -1437,9 +1424,153 @@ public String[] widgetLabelsForEdit() {
...
 				request.setParameter(widgets.get("Location:"), "2");
 				request.setParameter(widgets.get("Provider:"), "502");
+				//When: Html form is edited with an encounter date of today in which state X is selected
+				request.setParameter(widgets.get("Date:"), dateAsString(TODAY));
+				request.setParameter(widgets.get("State:"), START_STATE);
+			}
+		}.run();
+	}
+	
+	// See HTML-321
```
Comment (mseaton) [inline api-tests/src/test/java/org/openmrs/module/htmlformentry/WorkflowStateTagTest.java:1434]: This demonstrates the issue that was documented many years ago in HTML-321.  It fails as described without the fix.

File: api-tests/src/test/java/org/openmrs/module/htmlformentry/WorkflowStateTagTest.java [hunk 2]
```diff
@@ -1437,9 +1424,153 @@ public String[] widgetLabelsForEdit() {
...
+				results.assertNoErrors();
+				//Then: the existing enrollment with state X remains the same, starting one year ago
+				PatientProgram pp = assertProgram(results.getPatient(), TEST_PROGRAM, ONE_YEAR_AGO, null);
+				assertState(pp, START_STATE, ONE_YEAR_AGO, null);
+			}
+			
+		}.run();
+	}
+	
+	@Test
```
Comment (mseaton) [inline api-tests/src/test/java/org/openmrs/module/htmlformentry/WorkflowStateTagTest.java:1501]: This demonstrates the issue that was recently reported by @lnball ( or at least the first part of that issue that she documents ).

File: api-tests/src/test/java/org/openmrs/module/htmlformentry/WorkflowStateTagTest.java [hunk 3]
```diff
@@ -1838,18 +1969,31 @@ private void transitionToState(String state) {
 	}
 	
 	private void transitionToState(String state, Date date) {
+		transitionToState(state, date, null);
+	}
+	
+	private void transitionToState(String state, Date startDate, Date endDate) {
```
Comment (mseaton) [inline api-tests/src/test/java/org/openmrs/module/htmlformentry/WorkflowStateTagTest.java:1975]: This just modifies this utility method to make it possible to include an end date for the state when setting up a new patientprogram/state

File: api/src/main/java/org/openmrs/module/htmlformentry/FormEntrySession.java [hunk 4]
```diff
@@ -707,21 +707,33 @@ public void applyActions() throws BadFormDesignException {
 			Date encounterDate = encounter.getEncounterDatetime();
 			
 			// Try to determine if this is intended as an edit to an existing state
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/htmlformentry/FormEntrySession.java:709]: The code below is the fix.  Just as you described @mogoodrich , the appropriate fix here is not to edit an existing state unless the start date of that state matches the previously saved encounter date (on edit).

**Review by mogoodrich — APPROVED** [review]
LGTM, thanks @mseaton !

**Pull request merged**